breaking changes for Trixi.jl v0.5

ranocha opened the pull request:
This includes a first set of breaking changes for Trixi.jl v0.5 described in https://github.com/trixi-framework/Trixi.jl/issues/938.

@jlchan Could you please check that I removed only the `DGMulti` code that you intended to be removed?

I will leave this in draft mode until we are ready for the breaking release of Trixi.jl. We also need to update the compat bounds of Trixi.jl at Trixi2Vtk.jl to create our documentation here.

jlchan commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -249,7 +249,7 @@ function compute_flux_differencing_SBP_matrices(dg::DGMulti, sparse_operators)
   Qrst_hybridized, VhP, Ph = StartUpDG.hybridized_SBP_operators(rd)
   Qrst_skew = map(A -> 0.5 * (A - A'), Qrst_hybridized)
-  if sparse_operators isa Val{true}
+  if sparse_operators == true
```
Will this work? It seems like `Static.True` isn't the same as `true`?
```julia
julia> using Static
julia> True == true
false
```

ranocha replied:
We do not compare the types but instances of the type:
```julia
julia> using Static

julia> True() == true
true
```

jlchan replied:
Thanks!

jlchan commented on src/solvers/dgmulti/flux_differencing.jl:
```diff
@@ -261,7 +261,7 @@ function compute_flux_differencing_SBP_matrices(dg::DGMultiFluxDiffSBP, sparse_o
   Qrst = map(D -> M * D, Drst)
   Qrst_skew = map(A -> 0.5 * (A - A'), Qrst)
-  if sparse_operators isa Val{true}
+  if sparse_operators == true
```
Same as above

jlchan commented on src/solvers/dgmulti/types.jl:
```diff
@@ -143,39 +117,6 @@ GeometricTermsType(mesh_type::Curved, element_type::AbstractElemShape) = NonAffi
 # other potential constructor types to add later: Bilinear, Isoparametric{polydeg_geo}, Rational/Exact?
 # other potential mesh types to add later: Polynomial{polydeg_geo}?
 
-"""
-  DGMultiMesh(vertex_coordinates, EToV, dg::DGMulti{NDIMS};
-              is_on_boundary=nothing,
-              periodicity=ntuple(_->false, NDIMS)) where {NDIMS, Tv}
-
-- `vertex_coordinates` is a tuple of vectors containing x,y,... components of the vertex coordinates
-- `EToV` is a 2D array containing element-to-vertex connectivities for each element
-- `dg::DGMulti` contains information associated with to the reference element (e.g., quadrature,
-  basis evaluation, differentiation, etc).
-- `is_on_boundary` specifies boundary using a `Dict{Symbol, <:Function}`
-- `periodicity` is a tuple of booleans specifying if the domain is periodic `true`/`false` in the
-  (x,y,z) direction.
-"""
-# TODO: DGMulti v0.5. Standardize order of arguments, pass in `dg` first
-function DGMultiMesh(vertex_coordinates, EToV, dg::DGMulti{NDIMS};
-                     is_on_boundary=nothing,
-                     periodicity=ntuple(_->false, NDIMS), kwargs...) where {NDIMS}
-  if haskey(kwargs, :is_periodic)
-    # TODO: DGMulti, v0.5. Remove deprecated keyword
-    Base.depwarn("keyword argument `is_periodic` is now `periodicity`.", :DGMultiMesh)
-    periodicity=kwargs[:is_periodic]
-  end
-
-  md = MeshData(vertex_coordinates, EToV, dg.basis)
-  if NDIMS == 1
-    md = StartUpDG.make_periodic(md, periodicity...)
-  else
-    md = StartUpDG.make_periodic(md, periodicity)
-  end
-  boundary_faces = StartUpDG.tag_boundary_faces(md, is_on_boundary)
-  return DGMultiMesh(dg, GeometricTermsType(VertexMapped(), dg), md, boundary_faces)
-end
-
```
I'm not sure we want to remove this - there are a few constructors for `DGMultiMesh`. The `cells_per_dimension` version creates a uniform mesh; this one allows for general unstructured mesh construction.

ranocha replied:
I see, thanks. I added it back again in 2992733ac

jlchan reviewed (changes requested):
@ranocha thanks for organizing this. 

For `DGMulti`, it would be nice to make the `DGMultiMesh` constructors consistent so that `dg` is the first argument. This would be a breaking change. If there is some time before v0.5, I can take care of that sometime next week.

ranocha commented:
> @ranocha thanks for organizing this.
> 
> For `DGMulti`, it would be nice to make the `DGMultiMesh` constructors consistent so that `dg` is the first argument. This would be a breaking change. If there is some time before v0.5, I can take care of that sometime next week.

That sounds good to me - it would be great if you could introduce the corresponding deprecations next week :+1:

jlchan commented:
@ranocha I added the deprecations, along with some internal cleanup.

I removed a constructor in https://github.com/trixi-framework/Trixi.jl/pull/1302/commits/2a0205dad2b024dc779b997ff9233c24c32af435 which is, as far as I can tell, unused. I wasn't sure how to deprecate this properly - do we need to?

ranocha commented:
> I added the deprecations, along with some internal cleanup.

Thanks!

> I removed a constructor in [2a0205d](https://github.com/trixi-framework/Trixi.jl/commit/2a0205dad2b024dc779b997ff9233c24c32af435) which is, as far as I can tell, unused. I wasn't sure how to deprecate this properly - do we need to?

No, I think its okay. Was it documented, e.g., via a docstring? If so, it would be great if you could mention it in NEWS.md.

jlchan commented:
No docstring. I'll add it to the news file soon

ranocha commented:
Thanks!

sloede commented on src/Trixi.jl:
```diff
@@ -52,7 +52,7 @@ using P4est
 using RecipesBase: RecipesBase
 using Requires: @require
-using Static: Static, One
+using Static: Static, One, True, False
```
`using True, False` will automatically make someone who is `using Trixi` allow to use `Trixi.True()` and `Trixi.False()`?

ranocha replied:
Yes

sloede replied:
Thanks!

sloede commented on NEWS.md:
```diff
@@ -4,6 +4,30 @@ Trixi.jl follows the interpretation of [semantic versioning (semver)](https://ju
+  with nonconservative terms are created. Change
+  `Trixi.has_nonconservative_terms(::YourEquations) = Val{true}()` to
+  `Trixi.has_nonconservative_terms(::YourEquations) = Trixi.True()`.
+
```
```suggestion
- The (non-exported) DGSEM function `split_form_kernel!` has been renamed to `flux_differencing_kernel!`

```
I dunno if we should include it in the news, but since this has been a staple of Trixi for ~2 years, I thought it might be helpful to mention it here even though it is not part of the official API.

sloede reviewed (commented):
Two small comments, otherwise this LGTM!

sloede commented on NEWS.md:
```diff
@@ -4,6 +4,30 @@ Trixi.jl follows the interpretation of [semantic versioning (semver)](https://ju
+  with nonconservative terms are created. Change
+  `Trixi.has_nonconservative_terms(::YourEquations) = Val{true}()` to
+  `Trixi.has_nonconservative_terms(::YourEquations) = Trixi.True()`.
+- The (non-exported) DGSEM function `split_form_kernel!` has been renamed to `flux_differencing_kernel!`
```
```suggestion
- The (non-exported) DGSEM function `split_form_kernel!` has been renamed to `flux_differencing_kernel!`

```
Missing blank line.

sloede reviewed (approved):
🎉
